zebra-state/src/service/finalized_state/disk_format/transparent.rs

```diff
@@ -647,35 +766,69 @@ impl FromDisk for OutputLocation {
     }
 }
 
-impl IntoDisk for AddressBalanceLocation {
-    type Bytes = [u8; BALANCE_DISK_BYTES + OUTPUT_LOCATION_DISK_BYTES];
+impl<C: Constraint + Copy + std::fmt::Debug> IntoDisk for AddressBalanceLocationInner<C> {
+    type Bytes = [u8; BALANCE_DISK_BYTES + OUTPUT_LOCATION_DISK_BYTES + size_of::<u64>()];
 
     fn as_bytes(&self) -> Self::Bytes {
         let balance_bytes = self.balance().as_bytes().to_vec();
         let address_location_bytes = self.address_location().as_bytes().to_vec();
+        let received_bytes = self.received().to_le_bytes().to_vec();
 
-        [balance_bytes, address_location_bytes]
+        [balance_bytes, address_location_bytes, received_bytes]
             .concat()
             .try_into()
             .unwrap()
     }
 }
 
-impl FromDisk for AddressBalanceLocation {
+impl IntoDisk for AddressBalanceLocation {
+    type Bytes = [u8; BALANCE_DISK_BYTES + OUTPUT_LOCATION_DISK_BYTES + size_of::<u64>()];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.0.as_bytes()
+    }
+}
+
+impl IntoDisk for AddressBalanceLocationChange {
+    type Bytes = [u8; BALANCE_DISK_BYTES + OUTPUT_LOCATION_DISK_BYTES + size_of::<u64>()];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.0.as_bytes()
+    }
+}
+
+impl<C: Constraint + Copy + std::fmt::Debug> FromDisk for AddressBalanceLocationInner<C> {
     fn from_bytes(disk_bytes: impl AsRef<[u8]>) -> Self {
-        let (balance_bytes, address_location_bytes) =
-            disk_bytes.as_ref().split_at(BALANCE_DISK_BYTES);
+        let (balance_bytes, rest) = disk_bytes.as_ref().split_at(BALANCE_DISK_BYTES);
+        let (address_location_bytes, received_bytes) = rest.split_at(BALANCE_DISK_BYTES);
 
         let balance = Amount::from_bytes(balance_bytes.try_into().unwrap()).unwrap();
         let address_location = AddressLocation::from_bytes(address_location_bytes);
+        // # Backwards Compatibility
+        //
+        // If the value is missing a `received` field, default to 0.
+        let received = u64::from_le_bytes(received_bytes.try_into().unwrap_or_default());
```

Comment:
I think it would be nice to make this also forwards compatible, so that if `received_bytes` is larger than a u64 (i.e. a new field was added after it in a new database upgrade but zebra was downgraded) then this would still parse the value

Reply:
Fixed in https://github.com/ZcashFoundation/zebra/pull/9539/commits/e6c4f5e7867434e84728f04bbbfcbfc5ef00c957.